lib/Monad_WP/OptionMonad.thy

```diff
@@ -88,6 +112,18 @@ lemma opt_pred_disj:
   "((P1 |< hp) p \<or> (P2 |< hp) p) = (((P1 or P2) |< hp) p)"
   by (fastforce simp: pred_disj_def split: option.splits)
 
+lemma opt_predD:
+  "(P |< proj) x \<Longrightarrow> \<exists>y. proj x = Some y \<and> P y"
+  by (clarsimp split: option.splits)
+
+lemma opt_pred_unfold:
+  "(P |< f |> g)=  ((P |< g) |< f)"
+  by (fastforce simp: opt_map_def split: option.splits)
+
+lemma opt_pred_unfold_proj:
```

Comment:
This one actually unfolds, so that is fine. Not sure about the use site, but I'm tempted to add a lemma with `o` for this. Is this true `P |< f ||> g = P o g |< f`?

Reply:
> I'm tempted to add a lemma with `o` for this. Is this true `P |< f ||> g = P o g |< f`?

Ah, that should work, yes.